Import resiliency (1/3)

BenEmdon opened the pull request:
## What
This PR (1/3) proposes a different architecture to improve repository creation/import resiliency for the purpose of addressing #1253.
The architecture I am proposing is to move repository creation/import into a background job and communicate with the client asynchronously via ActionCable. Moving creation/import to a background job is a easy first step to resiliency since ActiveJobs can be retried.

I've decided to break this refactor into 3 parts:
1. Create `CreateGitHubRepositoryJob` job which is more resilient than the current offering
1. Create some form of porter polling job that can watch the status of the starter code import
1. Implement the new jobs in place of the current offering. Use ActionCable to asynchronously message the client with status.

## How
In this PR I created `CreateGitHubRepositoryJob` which essentially performs the same logic as [what is currently in place](https://github.com/education/classroom/blob/0638f37ce1ebc978926899cb8f64cc2b7fd4e8bf/app/models/assignment_repo/creator.rb#L58-L87). The big difference is that if at any point the repo creation/import fails, the GitHub repository is deleted and the creation/import job is retried. (Edit: the retry is done with exponential backoff).

Could I get some 👀 from @johndbritton or @tarebyte

BenEmdon commented on app/jobs/assignment_repo/create_github_repository_job.rb:
```diff
@@ -0,0 +1,47 @@
+class AssignmentRepo
+  class CreateGitHubRepositoryJob < ApplicationJob
+    queue_as :create_repository
+    retry_on Creator::Result::Error, queue: :create_repository
```
`retry_on` defaults to 3s wait and 5 attempts. 
This should be enough for what we are trying to do, right?

Also if we wanted we could pass a `wait: :exponentially_longer` which would perform an exponential backoff algorithm.

tarebyte replied:
I'd vote for the exponential backoff.

johndbritton replied:
Yeah, especially because the main reason we'd hit a problem like this is GitHub API downtime / issues.

BenEmdon replied:
I'll add 🚩

johndbritton commented on spec/jobs/create_github_repository_job_spec.rb:
```diff
@@ -0,0 +1,128 @@
+      result = assignment.assignment_repos.first
+      expect(result.nil?).to be_falsy
+      expect(result.assignment).to eql(assignment)
+      expect(result.user).to eql(student)
```
Style point on specs: Aim to have one `expect` per it block. If you need multiple `expects` split them up into separate it blocks and group them together logically in a `context`block.

BenEmdon commented:
## Update
* I added the exponential backoff 🚩
* I changed spec style to use shared contexts

## Response
> I would expect a side effect to happen if it fails and doesn't proceed to the next step. 
  @johndbritton 

When ActionCable is set up we can broadcast the progress or failure of the repo create/import. Having a progress bar would also be a nice user experience.
